src/spec.c

```diff
@@ -212,33 +214,52 @@ int isRdbLoading(RedisModuleCtx *ctx) {
   return isLoading == 1;
 }
 
-static void IndexSpec_TimedOutProc(RedisModuleCtx *ctx, IndexSpec *sp) {
+static void IndexSpec_TimedOutProc(RedisModuleCtx *ctx, WeakRef w_ref) {
   // we need to delete the spec from the specDict_g, as far as the user see it,
   // this spec was deleted and its memory will be freed in a background thread.
+
+  // attempt to promote the weak ref to a strong ref
+  StrongRef ref = WeakRef_Promote(w_ref);
+  WeakRef_Release(w_ref);
+
+  IndexSpec *sp = StrongRef_Get(ref);
+  if (!sp) {
+    // the spec was already deleted, nothing to do here
+    return;
+  }
 #ifdef _DEBUG
   RedisModule_Log(NULL, "notice", "Freeing index %s by timer", sp->name);
 #endif
 
   sp->isTimerSet = false;
+  // This function will perform an index drop, and we will still have to return our references
   IndexSpec_Free(sp);
 
+  StrongRef_Release(ref);
+
 #ifdef _DEBUG
   RedisModule_Log(NULL, "notice", "Freeing index by timer: done");
 #endif
 }
 
-static void IndexSpec_SetTimeoutTimer(IndexSpec *sp) {
+static void IndexSpec_SetTimeoutTimer(IndexSpec *sp, WeakRef ref) {
   if (sp->isTimerSet) {
-    RedisModule_StopTimer(RSDummyContext, sp->timerId, NULL);
+    WeakRef old_timer_ref;
+    if (RedisModule_StopTimer(RSDummyContext, sp->timerId, (void **)&old_timer_ref) == REDISMODULE_OK) {
+      WeakRef_Release(old_timer_ref);
+    }
   }
   sp->timerId = RedisModule_CreateTimer(RSDummyContext, sp->timeout,
-                                        (RedisModuleTimerProc)IndexSpec_TimedOutProc, sp);
+                                        (RedisModuleTimerProc)IndexSpec_TimedOutProc, *(void **)&ref);
```

Comment:
isn't `ref` gets out of scope?

Reply:
It is owned by the timer and will be released in the callback or if the timer is stopped, by the one who stopped it.